sv_pipeline/genome/utils/mapping_gene_ids.py

```diff
@@ -0,0 +1,106 @@
+
+            record = dict(zip(GENCODE_FILE_HEADER, fields))
+
+            if record['feature_type'] != 'gene':
+                continue
+
+            # parse info field
+            info_fields = [x.strip().split() for x in record['info'].split(';') if x != '']
+            info_fields = {k: v.strip('"') for k, v in info_fields}
+
+            gene_id_mapping[info_fields['gene_name']] = info_fields["gene_id"]
+
+    handle = open(pickle_file, 'wb')
+    pickle.dump(gene_id_mapping, handle, protocol=pickle.HIGHEST_PROTOCOL)
+    handle.close()
+
+
+# for performance comparison. It seems that hail search with filtering is slow. And converting hail table to Python JSON is also slow.
+def hail_load_gtf_data(gene_id_mapping, gencode_gtf_path):
+    gtf_table = hl.import_table(gencode_gtf_path, force=True, comment=['#'], no_header=True)
+    gtf_table = gtf_table.filter(gtf_table.f2 == 'gene')
+    info_fields = gtf_table.f8.split('; ').filter(lambda x: x.startswith('gene_id ') | x.startswith('gene_name ')).map(lambda x: x.split(' '))
+    gtf_table = gtf_table.annotate(gene_id_mapping=(info_fields[1][1].replace('"', ''),
+                                                    info_fields[0][1].replace('"', '')))
+    gene_id_mapping.update(hl.eval(hl.dict(gtf_table.gene_id_mapping.collect())))
+
+
+def load_gencode(gencode_release, gencode_gtf_path=None, genome_version=None, download_path=None):
+    """Update GeneInfo and TranscriptInfo tables.
+
+    Args:
+        gencode_release (int): the gencode release to load (eg. 25)
+        gencode_gtf_path (str): optional local file path of gencode GTF file. If not provided, it will be downloaded.
+        genome_version (str): '38'. Required only if gencode_gtf_path is specified.
+        download_path (str): The path for downloaded data
+    """
+    global gene_id_mapping
+    gene_id_mapping = {}
+
+    if gencode_gtf_path and genome_version and os.path.isfile(gencode_gtf_path):
+        if gencode_release <= 22:
+            raise Exception("Invalid genome_version: {}. Only gencode v23 and up is supported".format(genome_version))
+        elif genome_version != GENOME_VERSION_GRCh38:
+            raise Exception("Invalid genome_version for file: {}. Genome_version arg must be GRCh38".format(gencode_gtf_path))
+    elif gencode_gtf_path and not genome_version:
+        raise Exception("The genome version must also be specified after the gencode GTF file path")
+    else:
+        url = GENCODE_GTF_URL.format(gencode_release=gencode_release)
+        local_filename = download_file(url, to_dir=download_path)
+        gencode_gtf_path = local_filename
+        genome_version = GENOME_VERSION_GRCh38
+
+    logger.info("Loading {} (genome version: {})".format(gencode_gtf_path, genome_version))
+    load_gtf_data(gene_id_mapping, gencode_gtf_path)
+
+    logger.info('Get {} gene id mapping records'.format(len(gene_id_mapping)))
+    gene_id_mapping = hl.literal(gene_id_mapping)
```

Comment:
You are more than welcome to review other files. I just wanted to save your time for reviewing.

The `gene_id_mapping` is a global variable and used by `get_gene_id()` ar line #95. It is not a good coding style by using a global variable. I maybe should change them into a class.

Reply:
I agree that globals are the wrong choice here, but I also don't think a class is a better solution. You could just have this function just return a dictionary and use it. My comment on a previous PR review wasn't that you shouldn't return a dictionary, it was just that you should call this helper function in the function that needs it and not in the main function so the dictionary doesn't need to be passed around. To me, the  most readable way to do it is to load gencode when you are doing the annotation for the gene_affected intermediate field
```
gene_id_mapping = load_gencode(args.gencode_release, genome_version=WGS_SAMPLE_TYPE, download_path=args.gencode_path)

rows = rows.annotate(gene_affected=hl.filter(lambda x: hl.is_defined(x[0]),
            [(rows.info[col].map(lambda gene:  hl.struct(gene_symbol=gene, gene_id=gene_id_mapping.get(gene))), col.split('__')[-1])
            for col in  [gene_col for gene_col in rows.info if gene_col.startswith('PROTEIN_CODING__')
            and rows.info[gene_col].dtype == hl.dtype('array<str>')]]))
```

Reply:
I'd like to use `gene_id_mapping[gene]` rather than `gene_id_mapping.get(gene)` because it should be an error if a gene is not found in the mapping table. It shouldn't silently return an empty value in the gene missing case.